File: packages/runtime/ios/Engine/Detectors/Hook/HookSignatures.swift
import Foundation

/// Versioned signature data for iOS hook detection.
public enum HookSignatures {

  public static let version = "2026.08.1"

  /// Symbols checked against the image expected to provide them.
  ///
  /// Ordinary libSystem entry points. The question is not "is this symbol
  /// hooked" but "does this symbol still live in a system image" — a redirected
  /// `open` or `connect` resolving inside an injected library is a strong
  /// indicator, and one that renaming the library does not hide.
  public static let watchedSymbols = ["open", "read", "write", "connect", "fopen", "dlsym"]

  /// Classes and selectors whose implementation should come from a system image.
  ///
  /// Chosen because they are the ones instrumentation most often intercepts:
  /// network traffic and pasteboard access.
  public static let watchedMethods: [(className: String, selector: String)] = [
    (className: "NSURLSession", selector: "dataTaskWithRequest:completionHandler:"),
    (className: "NSURLSession", selector: "sharedSession"),
    (className: "UIPasteboard", selector: "generalPasteboard"),
  ]

  /// Absolute prefixes of operating-system image paths on a device.
  public static let systemPrefixes = ["/System/", "/usr/lib/"]

  /// Where the simulator nests those same prefixes.
  private static let simulatorRoot = "/RuntimeRoot"

  /// Whether an image path belongs to the operating system.
  ///
  /// Matched as a **prefix**, not a substring. A rootless jailbreak installs to
  /// `/var/jb/usr/lib/…`, which *contains* `/usr/lib/` — substring matching would
  /// classify every injected library on a modern jailbroken device as a system
  /// image and blind this check entirely, which is precisely where iOS hooks live
  /// today.
  ///
  /// The simulator nests the same prefixes under the Xcode runtime root, so that
  /// one specific nesting is accepted as well.
  public static func isSystemImage(_ path: String) -> Bool {
    if systemPrefixes.contains(where: { path.hasPrefix($0) }) {
      return true
    }
    return systemPrefixes.contains { path.contains(simulatorRoot + $0) }
  }
}

File: packages/runtime/android/src/main/java/com/rnsecurity/detectors/hook/HookSignatures.kt
package com.rnsecurity.detectors.hook

/**
 * Versioned signature data for hook and instrumentation detection.
 *
 * Everything here is a name an instrumentation framework happens to use today.
 * Names are the easiest thing in the world to change, which is precisely why
 * these signals are weighted `medium` and why the list is versioned rather than
 * scattered through the detectors.
 */
internal object HookSignatures {

  const val VERSION = "2026.08.1"

  /** Substrings seen in the memory map of a process under dynamic instrumentation. */
  val INSTRUMENTATION_MAP_MARKERS =
    listOf(
      "frida",
      "gadget",
      "gum-js-loop",
      "linjector",
      "libsubstrate",
      "substrate",
      "libwhale",
      "libdobby",
      "libepic"
    )

  /** Worker-thread names injected by instrumentation frameworks. */
  val INSTRUMENTATION_THREAD_NAMES =
    listOf("gmain", "gdbus", "gum-js-loop", "pool-frida", "pool-spawner", "linjector")

  /** Classes that exist only when a managed-code hooking framework is loaded. */
  val HOOK_FRAMEWORK_CLASSES =
    listOf(
      "de.robv.android.xposed.XposedBridge",
      "de.robv.android.xposed.XposedHelpers",
      "de.robv.android.xposed.IXposedHookLoadPackage",
      "org.lsposed.lspd.core.Main",
      "com.saurik.substrate.MS",
      "me.weishu.epic.art.Epic"
    )

  /** Package prefixes that should never appear on this application's call stack. */
  val HOOK_FRAMEWORK_STACK_PREFIXES =
    listOf("de.robv.android.xposed", "org.lsposed", "com.saurik.substrate", "me.weishu.epic")

  /** Memory-map substrings specific to managed-code hooking frameworks. */
  val HOOK_FRAMEWORK_MAP_MARKERS = listOf("lspd", "edxposed", "xposed", "riru", "libriru")

  /**
   * Native symbols checked against the library expected to provide them.
   *
   * Deliberately a short list of ordinary libc entry points. The check is not
   * "is this symbol hooked" but "does this symbol still live where it should" —
   * a redirected `open` or `connect` resolving inside an injected library is a
   * strong indicator, and one that renaming the library does not hide.
   */
  val EXPECTED_SYMBOL_LIBRARIES =
    mapOf(
      "open" to "libc.so",
      "read" to "libc.so",
      "write" to "libc.so",
      "connect" to "libc.so",
      "fopen" to "libc.so",
      "dlopen" to "libdl.so",
      "dlsym" to "libdl.so"
    )
}
